### PR TITLE
Few improvements

### PR DESCRIPTION
**Pull request by MBkkt**
### Purpose

[Few fixes for coroutine:](https://github.com/YACLib/YACLib/commit/a67f7b2870c70f1e7ec048984a3c39b68c477852) 

* ~~Add return_value(FutureBase<V, E>&& f), that allow to return value from future with same type as coroutine more free than co_return co_await f;~~
* Add return_value(StopTag) for cancel coroutine Future from function/Promise side
* Add ability to return error or result from coroutine Future
* Replace return_void with return_value(Unit) and co_return; with co_return {}; for allow use other return_value(...) in void coroutine

[Fix BaseCore possible bug with incorrect static_cast ...](https://github.com/YACLib/YACLib/commit/19f51745aaa214d2ef38930d7d8f64c7934ddb1c) 

[Few wait* fixes:](https://github.com/YACLib/YACLib/commit/9b3ca2ff8e9db866b4b17bfb12cb9aca4bfad447) 

* Remove WaitGroup and AwaitGroup duplication
* Make WaitGroup and OneShotEvent better
* Make Wait single Future faster

### Related Information

- [ ] Design document: ...
- [ ] Bench PR: ...

### Testing

- [ ] This change is a trivial rework or code cleanup without any test coverage.
- [ ] This change is already covered by existing tests.
- [ ] This PR adds tests that were used to verify all changes.
- [ ] There are tests in an external testing repository: ...

### CODE REVIEW

**Review by kononovk — DISMISSED** [review]

File: test/unit/coroutine/await_group.cpp [hunk 1]
```diff
@@ -85,10 +86,10 @@ TEST(AwaitGroup, OneWaiter) {
   EXPECT_TRUE(worker_done);
 }
 
-TEST(AwaitGroup, Workers) {
+TEST(WaitGroup, Workers) {
```
Comment (MBkkt) [inline test/unit/coroutine/await_group.cpp:89]: shit

File: test/unit/not_implemented.cpp [hunk 2]
```diff
@@ -85,10 +85,12 @@ TEST(CoroDummy, BaseCoroGetHandle) {
 #endif
 }
 
-TEST(AwaitGroupDummy, Cancel) {
+TEST(WaitGroupDummy, Cancel) {
```
Comment (MBkkt) [inline test/unit/not_implemented.cpp:88]: shit

File: test/unit/coroutine/await_group.cpp [hunk 3]
```diff
@@ -214,10 +218,10 @@ TEST(AwaitGroup, WithFutures) {
   scheduler->Wait();
 }
 
-TEST(AwaitGroup, SwichThread) {
+TEST(WaitGroup, SwichThread) {
```
Comment (MBkkt) [inline test/unit/coroutine/await_group.cpp:221]: shit

File: test/unit/coroutine/await_group.cpp [hunk 4]
```diff
@@ -176,10 +180,10 @@ TEST(AwaitGroup, BlockingWait) {
   scheduler->HardStop();
   scheduler->Wait();
 }
-TEST(AwaitGroup, WithFutures) {
+TEST(WaitGroup, WithFutures) {
```
Comment (MBkkt) [inline test/unit/coroutine/await_group.cpp:183]: shit

File: test/unit/coroutine/await_group.cpp [hunk 5]
```diff
@@ -134,38 +137,39 @@ TEST(AwaitGroup, Workers) {
   EXPECT_GE(steps, kWaiters + kWorkers * kYields);
 }
 
-TEST(AwaitGroup, BlockingWait) {
+TEST(WaitGroup, BlockingWait) {
```
Comment (MBkkt) [inline test/unit/coroutine/await_group.cpp:140]: shit

File: test/unit/coroutine/await_group.cpp [hunk 6]
```diff
@@ -45,26 +44,27 @@ TEST(AwaitGroup, JustWorks) {
   tp->Wait();
 }
 
-TEST(AwaitGroup, OneWaiter) {
+TEST(WaitGroup, OneWaiter) {
```
Comment (MBkkt) [inline test/unit/coroutine/await_group.cpp:47]: shit

File: test/unit/coroutine/await_group.cpp [hunk 7]
```diff
@@ -21,7 +21,7 @@ namespace {
 
 using namespace std::chrono_literals;
 
-TEST(AwaitGroup, JustWorks) {
+TEST(WaitGroup, JustWorks) {
```
Comment (MBkkt) [inline test/unit/coroutine/await_group.cpp:24]: shit

File: src/util/mutex_event.cpp [hunk 8]
```diff
@@ -12,20 +12,14 @@ void MutexEvent::Wait(Token& token) noexcept {
...
 
-void MutexEvent::Reset() noexcept {
-  _is_ready = false;
-}
-
-void MutexEvent::SetOne() noexcept {
+void MutexEvent::Set() noexcept {
   std::lock_guard lock{_m};
   _is_ready = true;
-  _cv.notify_one();  // Notify under mutex, because cv located on stack memory of other thread
```
Comment (MBkkt) [inline src/util/mutex_event.cpp:22]: notify_one

File: include/yaclib/coroutine/detail/await_group_impl.hpp [hunk 9]
```diff
@@ -3,7 +3,7 @@
 namespace yaclib {
```
Comment (MBkkt) [inline include/yaclib/coroutine/detail/await_group_impl.hpp:3]: remove file

**Pull request merged**